drivers/gpu/drm/rp1/rp1-vec/rp1_vec.c

```diff
@@ -231,36 +198,47 @@ static const struct drm_display_mode rp1vec_modes[4] = {
 	}
 };
 
+/*
+ * Advertise standard and preferred video modes.
+ *
+ * From each interlaced mode in the table above, derive a progressive one.
+ *
+ * This driver always support all 50Hz and 60Hz video modes, regardless
+ * of connector's tv_mode; nonstandard combinations generally default
+ * to PAL[-BDGHIKL] or NTSC[-M] depending on resolution and field-rate
+ * (except that "PAL" with 525/60 will be implemented as "PAL60").
+ */
+
 static int rp1vec_connector_get_modes(struct drm_connector *connector)
 {
-	struct rp1_vec *vec = container_of(connector, struct rp1_vec, connector);
-	bool ok525 = RP1VEC_TVSTD_SUPPORT_525(vec->tv_norm);
-	bool ok625 = RP1VEC_TVSTD_SUPPORT_625(vec->tv_norm);
 	int i, prog, n = 0;
+	int tvm = rp1vec_initial_tv_mode(connector);
+	bool prefer625 =
+		tvm == DRM_MODE_TV_MODE_PAL   ||
+		tvm == DRM_MODE_TV_MODE_PAL_N ||
+		tvm == DRM_MODE_TV_MODE_SECAM;
 
 	for (i = 0; i < ARRAY_SIZE(rp1vec_modes); i++) {
-		if ((rp1vec_modes[i].vtotal == 625) ? ok625 : ok525) {
-			for (prog = 0; prog < 2; prog++) {
-				struct drm_display_mode *mode =
-					drm_mode_duplicate(connector->dev,
-							   &rp1vec_modes[i]);
-
-				if (prog) {
-					mode->flags &= ~DRM_MODE_FLAG_INTERLACE;
-					mode->vdisplay	  >>= 1;
-					mode->vsync_start >>= 1;
-					mode->vsync_end	  >>= 1;
-					mode->vtotal	  >>= 1;
-				}
-
-				if (mode->hdisplay == 704 &&
-				    mode->vtotal == ((ok525) ? 525 : 625))
-					mode->type |= DRM_MODE_TYPE_PREFERRED;
-
-				drm_mode_set_name(mode);
-				drm_mode_probed_add(connector, mode);
-				n++;
+		for (prog = 0; prog < 2; prog++) {
```

Comment:
The default timings for the defined TV standards can added with
```
	int count = drm_connector_helper_tv_get_modes(connector);
```
vc4 does this in the 6.6 tree.

Adding the cropped modes is a little trickier and may still require modes to be stored, but it's nicer to remove as much duplication as possible.

Reply:
Gosh there's a lot of code under `drm_connector_helper_tv_modes()`. Hope it's correct.
Will it produce both 625/50 and 625/50 modes at the same time?

Reply:
Maxime wrote it

It'll add 480i and/or 576i based on the standards listed `drm_mode_create_tv_properties` was told were supported, and in the order (+ preferred flag) as based on the default value registered when you `drm_object_attach_property` the tv_mode property.

Reply:
Ah, so that's what "default property value" means.

Reply:
Currently marking one of the cropped modes as "preferred": this is to stop the FB console spilling off the screen.